pip/pip-322.md

```diff
@@ -0,0 +1,143 @@
+throttling and added latency.
+
+Overhauling the built-in implementation will improve multi-tenancy,
+allow Pulsar to scale to demanding workloads, and address longstanding
+issues.
+
+Rate limiters act as a conduit to more extensive capacity management and
+Quality of Service (QoS) controls in Pulsar. They are integral to
+Pulsar's core multi-tenancy features. This refactoring will pave the way
+for future enhancements.
+
+# Background Reading
+
+To understand what we're trying to achieve, you should first read the
+blog post "Proposal for Pulsar Rate Limiting Enhancements"
+[here](https://codingthestreams.com/pulsar/2023/11/22/pulsar-slos-and-rate-limiting.html#proposal-for-pulsar-rate-limiting-enhancements).
+Please also make sure to read the [Pulsar Community Meeting minutes
+  2023/11/23](https://lists.apache.org/thread/y1sqpyv37fo0k4bm1ox28wggvkb7pbtw).
+
+# Goals
+
+## In Scope
+
+- Preserve current functionality without breaking changes
+- Consolidate the multiple existing rate limiting options into a single,
+  configurable rate limiting solution
+- Remove the separate “precise” rate limiter
+
+### Problems to Address  
+
+- High CPU load with default rate limiter
+- High lock contention that impacts shared Netty IO threads and adds
+  latency to unrelated Pulsar topic producers ([Bug] RateLimiter lock
+  contention when use precise publish rate limiter #21442.)
+- Multiple limiting implementations (default, precise) which
+  unnecessarily expose implementation details to users of Pulsar and
+  make the code harder to maintain and improve
+- Inability to limit throughput consistently when using default rate
+  limiter
+- Inconsistent behavior across multiple levels of throttling (broker,
+  namespace, connection)
+- Code maintainability
+  - Improve understandability of code
+
+## Out of Scope
+
+- Custom/pluggable rate limiters
+- Additional rate limiting features
+- Cluster-wide capacity management
+- Addressing the shared connection multiplexing problem where throttling
+  multiple independent streams multiplexed on the same connection cannot
+  be consistently throttled by pausing reads on the server side.
+
+# High Level Design
+
+The proposed refactor will overhaul rate limiting internals while
+preserving existing APIs and user-facing behavior. A token bucket
+algorithm will provide efficient and accurate calculations to throttle
+throughput. 
+
+Multiple built-in options such as "precise" rate limiter will be
+consolidated under a single solution. Performance issues caused by
+contention and CPU overhead will be addressed.
+
+## Publisher rate limiter throttling
+
+The refactored design ensures that a publish rate limiter operates in a
+non-blocking fashion. It consumes the quota seamlessly without
+disrupting ongoing operations. When the quota is exhausted, the
+connection is throttled by pausing reads.
+
+Subsequently, the publisher rate limiter adds the throttled producer to
```

Comment:
Not sure I understand this. When you say "publishers" in plural, you mean all publishers that share a single connection? If each is to a different topic, why unthrottle all of them when this task is executed? I mean, if one producer is throttled, then all producers are throttled?

Reply:
I'll rename "publishers" with "producers" to make it more clear. A single rate limiter applies to multiple producers in many cases. Publisher rate limiters can be at multiple levels: at namespace level, at broker level and at resource group level. 

> if one producer is throttled, then all producers are throttled?

I'll also need to explain it better that one of the key changes in the design is that throttling for each producer is applied only after a producer has sent message(s) to the broker. After the limit has been exceed, each producer will be able to send one more send command and the throttling happens after that. The result of this is similar as when everything would be throttled at once since the message rate is always an average over some period of time. The benefit is that idle producers don't need to be iterated and toggled when the rate limiter state is to throttle. This results in more efficient operations and there's also a better way to ensure fairness when throttling is applied only when there is new messages. When the publisher rate limiter is throttled, throttled producers will be queued and each one will get an even chance to produce new messages.

Reply:
I'll summarize the key points we discussed over Slack to streamline the explanation:
* Rate limits are defined in different scopes: topic, namespace, resource groups.
   Hence there is a PublisherRateLimiter per topic, namespace and resource group.
*There is a trade-off being taken by this PIP design. The current rate limiter is eager - once a producer crosses the rate limit, it iterates over all existing scope producers (topic/namespace/resource-group) and disabled the Netty ready flag, which effectively causes Netty to stop reading from those connections == back-pressure is applied and held at the client side. 
   In the new suggested rate limiter, works in a lazy fashion. When a Send command is received, only then the rate limit is checked. It allows this message to be processed in any case (since there is no way in the protocol to reject a message), and only then if the limit was crossed, it disables the Netty read flag (conceptually, the PIP design should be augmented to explain exactly what happens, as it involves several methods of Async Token Bucket, a Throttler, a delay task and an un-throttle producers queue). Hence lazy, since it lazily checks the limit per producer and not eagerly shuts down all producers when the rate limit for say a topic was crossed.
* When the producer receive a message, it first consumes the tokens from the async-token-bucket and only then checks if the limit was crossed. If the limit was crossed, it disables the Netty read flag for the connection through something called Throttler, and and places the producer into an un-throttle queue. A scheduled task is placed to un-throttle all throttled producers (on the queue) based on the delay calculated by the async-token-bucket.
* A Throttler is mechanism that helps manage several rate limiters for same scope (e.g. topic). A single connection (producer) may trip over more one rate limit scope: topic, namespace, resource group.
* Un-throttling from a queue guarantees fairness somehow.

Reply:
Thanks, I hopefully got this included in the most recent version of the PIP documented.